File: algo/algoevening/js/MergeSort.js
////////////////////////////////////////////////////////////
let one = [1, 3, 5, 6];
let two = [2, 3, 4];

let res = merge(one, two);
console.log(`Array is ${res}`);

function merge(one, two) {
    let result = [];
    let oneIndex = 0;
    let twoIndex = 0;

    while(oneIndex < one.length && twoIndex < two.length) {
        let oneElement = one[oneIndex];
        let twoElement = two[twoIndex];
        if(oneElement < twoElement) {
            result.push(oneElement);
            oneIndex++;
        } else {
            result.push(twoElement);
            twoIndex++;
        }
    }
    while(oneIndex < one.length){
        result.push(one[oneIndex]);
        oneIndex++;
    }
    while(twoIndex < two.length) {
        result.push(two[twoIndex]);
        twoIndex++;
    }
    return result;
}

let arr = [38,27,43,3,9,82,10];
let sortedArray = mergeSort(arr);
console.log(`Sorted array is ${sortedArray}`);

function mergeSort(arr) { ////38,27,43,3  // 38 27 // 38
    if(arr.length === 1) {
        return arr;
    }

    //находим середину нашего массива
    let mid = Math.floor(arr.length / 2);
    // создать два новых массива с (0 до mid - 1) и с (mid до array.length - 1)
    let left = arr.slice(0, mid); //38,27,43,3  // 38 27 // 38 // 43
    let right = arr.slice(mid); // 9,82,10    //  43 3 // 27  // 3

    let resLeft = mergeSort(left); //38,27,43,3 // 38 27 // 38 // 43
    // 38 // 27 38 // 43

    let resRight = mergeSort(right);// 27 // 43 3 // 3
    // 27 //3

    return merge(resLeft,resRight); // 27 38 ->

}

File: algo/algoevening/java/MergeSort.java
package algoevening.java;

import java.util.Arrays;

public class MergeSort {

    public static void main(String[] args) {
        int[] one = {1, 3, 5, 6};
        int[] two = {2, 3, 4};

        System.out.println(Arrays.toString(merge(one,two)));

        int[] arr = {38,27,43,3,9,82,10};
        System.out.println(Arrays.toString(mergeSort(arr)));

    }

    public static int[] mergeSort(int[] arr) {
        if (arr.length == 1) {
            return arr;
        }

        int mid = arr.length / 2;
        int[] left = Arrays.copyOfRange(arr, 0, mid);
        int[] right = Arrays.copyOfRange(arr, mid, arr.length);

        int[] resLeft = mergeSort(left);
        int[] resRight = mergeSort(right);

        return merge(resLeft, resRight);
    }

    public static int[] merge(int[] one, int[] two) {
        int oneIndex = 0;
        int twoIndex = 0;
        int resultIndex = 0;
        int[] result = new int[one.length + two.length];
        while (oneIndex < one.length && twoIndex < two.length) {
            int oneElement = one[oneIndex];
            int twoElement = two[twoIndex];
            if (oneElement < twoElement) {
                result[resultIndex] = oneElement;
                oneIndex++;
            } else {
                result[resultIndex] = twoElement;
                twoIndex++;
            }
            resultIndex++;
        }

        while(oneIndex < one.length) {
            result[resultIndex] = one[oneIndex];
            oneIndex++;
            resultIndex++;
        }

        while(twoIndex < two.length) {
            result[resultIndex] = two[twoIndex];
            twoIndex++;
            resultIndex++;
        }

        return result;
    }
}
